design/proposals/spike-cli-dockerless.md

```diff
@@ -0,0 +1,257 @@
+
+None of our CLIs test against other platforms besides Linux.
+
+|            | Security |
+| ---------- | -------- |
+| Ruby CLI   | See scan |
+| Go CLI     | See scan |
+| Python CLI | See scan |
+
+#### CLI commands to be supported
+
+The CLI should support the following commands:
+
+*Core* -  `init`, `authn` (`login`, `authenticate`),  `policy (load)`, `list`, `show `, `rotate_api_key` (`host` / `user`) and `fetch secrets`
+
+*Additional we might want to consider* - `resource`, `role`, `check`, `help`
+
+#### Estimation
+
+The following table summarizes the findings criteria detailed above and translates them to numbered equivalents. 
+
+1 - the least work
+
+3 - the most work
+
+|            | Functionality | Quality | UX   | Deployment | Testing | Security |
+| ---------- | ------------- | ------- | ---- | ---------- | ------- | -------- |
+| Ruby CLI   | 1             | 1       | 2    | 3          | 1       | See scan |
+| Go CLI     | 3             | 3       | 3    | 2          | 3       | See scan |
+| Python CLI | 2             | 2       | 1    | 1          | 2       | See scan |
+
+Ruby CLI ~11 weeks
+
+- Ramp-up on creating packages and create design for CLI (~1 week)
+- Provide package (rpm, pkg, deb, etc) / [deliver as executable](https://www.ruby-toolbox.com/categories/packaging_to_executables) (CLI + dependencies) (~3 week)
+- Expand current testing if needed (~1 week)
+- Build infrastructure for testing on new platforms (Windows, macOS, REHL environments) (~3 weeks)
+- Run a few CLI commands on new supported platforms (Windows, macOS environments) (~2 weeks)
+- Close open issues (~1 week)
+
+Python CLI ~13 weeks
+
+- Python ramp-up and create design for CLI (~1 week)
+- Implement logic to fill in functionality gaps (7 CLI commands, ~3 weeks)
+  - `init`, `authn` (`login`, `authenticate`), `show `, `rotate_api_key` (`host` / `user`)
+- Write associated tests for new implementation (~3 weeks)
+- Build infrastructure for testing on new platforms (Windows, macOS environments) (~3 weeks)
+- Run a few of them on different supported platforms (Windows, macOS, REHL environments) (~2 weeks)
+- Close open issues (~1 week)
+- Add documentation
+
+Go CLI - Complete rewrite (~18+ weeks)
+
+- Create design for CLI
+- CLI Ramp up
+- Redesign CLI
+- Implement new CLI commands and logic
+- Write tests for new functionality
+- Might need to redesign and build new infrastructure for environment testing (Windows, macOS, REHL environments)
+- Run a few CLI commands on new supported platforms (Windows, macOS, REHL environments)
+- Add documentation
+
+#### Takeways and recommendation
+
+Ruby requires the least amount of work in terms of what is already fuctionally supported but its UX is not intuitive. As part of its deployment, the CLI requires a specific version of Ruby installed on the machine to run the CLI.
+
+Machines may not have internet access so we will need to pre-package Ruby and other dependencies and deliver as with an installer.
```

Comment:
We'd need to package the gem and dependencies, but Ruby should already be available on the underlying system in order to run Conjur, Conjur-UI, and Evoke.

Reply:
> This pre-req also means that we cannot use the CLI on machines that do not have our exact Ruby requirement so effectively we will require a full VM for the CLI if we can't use Docker

I would argue that this is same for Python don't you think? We would ideally be packing Python and all dependencies needed to run the Python CLI

Reply:
> We'd need to use the system-level Ruby that's global (unless we work on developing some RVM solution) so all Ruby apps of the system must match in their Ruby requirement

@sgnn7  but this is a challenge we are currently facing with our other dockerless components (conjur, conjur ui, evoke) so it seems that this isn't new. 

@rafis3 correct me if I am wrong here but this is something we need to worry about regardless no?

Reply:
@jvanderhoof this document only discusses the Conjur CLI, so Conjur, UI and evoke are not relevant here.
@sgnn7 why do we have to use a system installed Ruby? Why won't we come with our own Ruby runtime and point the CLI to using that?

Reply:
> @sgnn7 but this is a challenge we are currently facing with our other dockerless components (conjur, conjur ui, evoke) so it seems that this isn't new

@sigalsax When talking about the appliance, ensuring system-level Ruby version is not a big deal as we must have our appliance in our own OS container/VM. The problem here is when we need a CLI that will run outside of the deployment infrastructure with the same requirement. If a user wanted to run the CLI but they needed to install a specific (and probably old) Ruby version for the whole machine, they would most likely have to dedicate a whole machine/VM/container for it if they didn't want a system that they cannot upgrade without problems.

> @sgnn7 why do we have to use a system installed Ruby? Why won't we come with our own Ruby runtime and point the CLI to using that?

Because we don't have a good Ruby isolation framework (e.g.`rvm`/`rbenv`) explored/used for the CLI (nor other components for that matter) so you must rely on the system-level Ruby version to provide your interpreter. It's not an unfixable problem but it's a really challenging one from a deployment standpoint.